File: packages/protocol/src/revocation.ts
/**
 * Cryptographically signed Revocation Records & mesh revocation sync.
 * Matches Go `packages/wire/revocation.go` byte-for-byte.
 */

import { bytesToHex, hexToBytes, utf8 } from './bytes.js';
import {
  type DeviceIdentity,
  deriveDeviceId,
  signDeviceMessage,
  validateDeviceId,
  verifyDeviceSignature,
} from './identity.js';

export const DOMAIN_REVOCATION_RECORD = 'sendbeam/2 revocation-record:';
export const MAX_REVOCATION_TIMESTAMP_SKEW_MS = 5 * 60 * 1000; // 5 minutes

export interface RevocationRecord {
  revoker_device_id: string;
  revoked_device_id: string;
  seq: number;
  timestamp: string; // RFC 3339 UTC format
  signature: string; // 128-char lowercase hex (64-byte Ed25519 signature)
}

/**
 * Build the canonical binary challenge for a revocation record.
 * Format: DomainRevocationRecord || RevokerDeviceID || RevokedDeviceID || BigEndian(Seq) || Timestamp
 */
export function buildRevocationChallenge(
  revokerId: string,
  revokedId: string,
  seq: number,
  timestamp: string,
): Uint8Array {
  const domainBytes = utf8(DOMAIN_REVOCATION_RECORD);
  const revokerBytes = utf8(revokerId);
  const revokedBytes = utf8(revokedId);

  const seqBytes = new Uint8Array(8);
  const view = new DataView(seqBytes.buffer);
  view.setBigUint64(0, BigInt(seq), false); // Big-endian

  const tsBytes = utf8(timestamp);

  const totalLen =
    domainBytes.length +
    revokerBytes.length +
    revokedBytes.length +
    seqBytes.length +
    tsBytes.length;
  const buf = new Uint8Array(totalLen);

  let offset = 0;
  buf.set(domainBytes, offset);
  offset += domainBytes.length;
  buf.set(revokerBytes, offset);
  offset += revokerBytes.length;
  buf.set(revokedBytes, offset);
  offset += revokedBytes.length;
  buf.set(seqBytes, offset);
  offset += seqBytes.length;
  buf.set(tsBytes, offset);

  return buf;
}

/**
 * Sign a new RevocationRecord using the local device's private identity key.
 */
export async function signRevocation(
  identity: DeviceIdentity,
  revokedDeviceId: string,
  seq: number,
  now = new Date(),
): Promise<RevocationRecord> {
  if (!identity) throw new Error('invalid identity: null or undefined');
  if (!validateDeviceId(revokedDeviceId)) {
    throw new Error(`invalid revoked device id: ${revokedDeviceId}`);
  }
  if (!Number.isInteger(seq) || seq <= 0) {
    throw new Error('seq must be a positive integer > 0');
  }

  const timestamp = now.toISOString();
  const challenge = buildRevocationChallenge(identity.deviceId, revokedDeviceId, seq, timestamp);
  const sigBytes = signDeviceMessage(identity, challenge);

  return {
    revoker_device_id: identity.deviceId,
    revoked_device_id: revokedDeviceId,
    seq,
    timestamp,
    signature: bytesToHex(sigBytes),
  };
}

/**
 * Sign a new self-tombstone RevocationRecord announcing this device's own revocation (ADR 0010 §4.2).
 */
export async function signSelfTombstone(
  identity: DeviceIdentity,
  seq: number,
  now = new Date(),
): Promise<RevocationRecord> {
  if (!identity) throw new Error('invalid identity: null or undefined');
  return signRevocation(identity, identity.deviceId, seq, now);
}

/**
 * Returns true if the revocation record is a self-tombstone where the revoker revokes itself (ADR 0010 §4.2).
 */
export function isSelfTombstone(record: RevocationRecord): boolean {
  return (
    !!record && !!record.revoker_device_id && record.revoker_device_id === record.revoked_device_id
  );
}

/**
 * Validate the structural integrity of a RevocationRecord.
 * Under ADR 0010 §4.2, self-tombstones (revoker_device_id === revoked_device_id) are structurally valid.
 */
export function validateRevocationRecord(record: RevocationRecord): void {
  if (!record) throw new Error('invalid revocation record: null or undefined');
  if (!validateDeviceId(record.revoker_device_id)) {
    throw new Error(`invalid revoker device id: ${record.revoker_device_id}`);
  }
  if (!validateDeviceId(record.revoked_device_id)) {
    throw new Error(`invalid revoked device id: ${record.revoked_device_id}`);
  }
  if (!Number.isInteger(record.seq) || record.seq <= 0) {
    throw new Error('seq must be a positive integer > 0');
  }
  if (!record.timestamp || isNaN(Date.parse(record.timestamp))) {
    throw new Error(`invalid timestamp: ${record.timestamp}`);
  }
  const sigBytes = hexToBytes(record.signature);
  if (sigBytes.length !== 64) {
    throw new Error('signature must be 64 bytes hex');
  }
}

/**
 * Verify a RevocationRecord against the revoker's public key and clock constraints.
 */
export async function verifyRevocation(
  record: RevocationRecord,
  revokerPublicKey: Uint8Array,
  maxSkewMs = MAX_REVOCATION_TIMESTAMP_SKEW_MS,
  now = new Date(),
): Promise<boolean> {
  try {
    validateRevocationRecord(record);
  } catch {
    return false;
  }

  const expectedRevokerId = await deriveDeviceId(revokerPublicKey);
  if (expectedRevokerId !== record.revoker_device_id) {
    return false;
  }

  const recordTime = Date.parse(record.timestamp);
  if (isNaN(recordTime)) return false;

  if (maxSkewMs > 0) {
    const skew = now.getTime() - recordTime;
    if (skew < -maxSkewMs) {
      // Timestamp is in the future beyond acceptable skew
      return false;
    }
  }

  const sigBytes = hexToBytes(record.signature);
  if (sigBytes.length !== 64) return false;

  const challenge = buildRevocationChallenge(
    record.revoker_device_id,
    record.revoked_device_id,
    record.seq,
    record.timestamp,
  );

  return verifyDeviceSignature(revokerPublicKey, challenge, sigBytes);
}

import {
  ERR_REVOCATION_SEQ_ROLLBACK,
  ERR_REVOCATION_UNAUTHORIZED,
  ERR_UNKNOWN_REVOKER,
} from './errors.js';
import type { TombstoneStore } from './tombstone-store.js';
import {
  RELATIONSHIP_CLUSTER_MEMBER,
  RELATIONSHIP_CLUSTER_OWNER,
  type TrustStore,
} from './trust-store.js';

export interface IngestRevocationOptions {
  tombstoneStore?: TombstoneStore;
  localClusterId?: string;
  localDeviceId?: string;
  localPublicKey?: Uint8Array;
  now?: Date;
  onAbortSession?: (deviceId: string) => void;
}

/**
 * Ingest and verify an incoming mesh RevocationRecord in accordance with ADR 0010 §4.4.
 */
export async function ingestRevocationRecord(
  record: RevocationRecord,
  trustStore: TrustStore,
  options?: IngestRevocationOptions,
): Promise<void> {
  // 1. Syntax validation
  validateRevocationRecord(record);

  const now = options?.now ?? new Date();
  const isSelf = record.revoker_device_id === record.revoked_device_id;
  const isLocal = Boolean(
    options?.localDeviceId && record.revoker_device_id === options.localDeviceId,
  );

  // 2. Retrieve Revoker A from local trust store
  let revokerPubBytes: Uint8Array;
  if (isLocal && options?.localPublicKey) {
    revokerPubBytes = options.localPublicKey;
  } else {
    const revoker = await trustStore.getDevice(record.revoker_device_id);
    if (!revoker) {
      if (isSelf) {
        if (options?.tombstoneStore) {
          await options.tombstoneStore.storeTombstone(record);
        }
        return;
      }
      throw new Error(ERR_UNKNOWN_REVOKER);
    }

    // 3. Check if Revoker A is already revoked locally
    if (revoker.revoked) {
      throw new Error('revocation revoker is revoked');
    }

    // 4. Authorization check
    if (!isSelf && !isLocal) {
      if (
        revoker.relationship !== RELATIONSHIP_CLUSTER_MEMBER &&
        revoker.relationship !== RELATIONSHIP_CLUSTER_OWNER
      ) {
        throw new Error(ERR_REVOCATION_UNAUTHORIZED);
      }
      if (!revoker.clusterId || revoker.clusterId.trim() === '') {
        throw new Error(ERR_REVOCATION_UNAUTHORIZED);
      }
      if (options?.localClusterId && revoker.clusterId !== options.localClusterId) {
        throw new Error(ERR_REVOCATION_UNAUTHORIZED);
      }
    }
    revokerPubBytes = hexToBytes(revoker.publicKey);
  }

  // 5. Retrieve Target B from local trust store
  const target = await trustStore.getDevice(record.revoked_device_id);
  if (!target) {
    if (options?.tombstoneStore) {
      await options.tombstoneStore.storeTombstone(record);
    }
    return;
  }

  // 6. Verify Ed25519 signature
  const sigValid = await verifyRevocation(
    record,
    revokerPubBytes,
    MAX_REVOCATION_TIMESTAMP_SKEW_MS,
    now,
  );
  if (!sigValid) {
    throw new Error('trusted-session signature verification failed');
  }

  // 7. Sequence monotonicity check
  if (target.revoked && target.revocationSeq && record.seq <= target.revocationSeq) {
    throw new Error(ERR_REVOCATION_SEQ_ROLLBACK);
  }

  // 9. Apply revocation to Target B in trust store
  await trustStore.revokeDeviceWithRecord(record);

  if (options?.tombstoneStore) {
    await options.tombstoneStore.storeTombstone(record);
  }

  // 10. Abort active sessions
  options?.onAbortSession?.(record.revoked_device_id);
}

File: packages/wire/revocation.go
package wire

import (
	"crypto/ed25519"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"fmt"
	"time"
)

// Revocation-related constants and domain separation (V17-PR01).
const (
	DomainRevocationRecord     = "sendbeam/2 revocation-record:"
	MaxRevocationTimestampSkew = 5 * time.Minute
)

var (
	// ErrInvalidRevocationRecord indicates a malformed or structurally invalid revocation record.
	ErrInvalidRevocationRecord = Errorf(CodeProtocol, "invalid revocation record")

	// ErrRevocationSignatureFailed indicates that the Ed25519 signature in a revocation record failed verification.
	ErrRevocationSignatureFailed = Errorf(CodeAuth, "revocation signature verification failed")

	// ErrRevocationSelfRevoke indicates a revocation record where the revoker and revoked device IDs are identical (deprecated: self-tombstones are valid under ADR 0010).
	ErrRevocationSelfRevoke = errors.New("cannot revoke self in mesh sync")

	// ErrRevocationSeqRollback indicates a revocation sequence number that is less than or equal to an existing record.
	ErrRevocationSeqRollback = Errorf(CodeProtocol, "revocation sequence number rollback")

	// ErrRevocationTimestampSkew indicates that the revocation timestamp is outside acceptable clock skew boundaries.
	ErrRevocationTimestampSkew = Errorf(CodeAuth, "revocation timestamp outside acceptable window")

	// ErrRevokerUntrusted indicates that the device submitting a revocation is not an active trusted peer.
	ErrRevokerUntrusted = Errorf(CodeAuth, "revocation revoker is not an active trusted peer")
)

// ValidateRevocationSeq asserts that incomingSeq is strictly greater than existingSeq.
func ValidateRevocationSeq(existingSeq, incomingSeq uint64) error {
	if incomingSeq <= existingSeq {
		return ErrRevocationSeqRollback
	}
	return nil
}

// RevocationRecord represents a cryptographically signed statement asserting that a peer device has been revoked.
type RevocationRecord struct {
	RevokerDeviceID string `json:"revoker_device_id"`
	RevokedDeviceID string `json:"revoked_device_id"`
	Seq             uint64 `json:"seq"`
	Timestamp       string `json:"timestamp"` // RFC 3339 UTC format
	Signature       string `json:"signature"` // hex-encoded Ed25519 signature
}

// IsSelfTombstone returns true if the revocation record is a self-tombstone where the revoker revokes itself (ADR 0010 §4.2).
func (r *RevocationRecord) IsSelfTombstone() bool {
	if r == nil {
		return false
	}
	return r.RevokerDeviceID != "" && r.RevokerDeviceID == r.RevokedDeviceID
}

// BuildRevocationChallenge constructs the canonical binary payload signed by the revoker device.
// Format: DomainRevocationRecord || RevokerDeviceID || RevokedDeviceID || BigEndian(Seq) || Timestamp
func BuildRevocationChallenge(revokerID, revokedID string, seq uint64, timestamp string) []byte {
	seqBytes := make([]byte, 8)
	binary.BigEndian.PutUint64(seqBytes, seq)

	buf := make([]byte, 0, len(DomainRevocationRecord)+len(revokerID)+len(revokedID)+8+len(timestamp))
	buf = append(buf, DomainRevocationRecord...)
	buf = append(buf, revokerID...)
	buf = append(buf, revokedID...)
	buf = append(buf, seqBytes...)
	buf = append(buf, timestamp...)
	return buf
}

// SignRevocation creates and signs a new RevocationRecord with the revoker's DeviceIdentity.
// Under ADR 0010 §4.2, RevokerDeviceID == RevokedDeviceID is permitted and creates an authorized self-tombstone.
func SignRevocation(id *DeviceIdentity, revokedDeviceID string, seq uint64, now time.Time) (*RevocationRecord, error) {
	if id == nil {
		return nil, ErrInvalidIdentity
	}
	if !ValidateDeviceID(revokedDeviceID) {
		return nil, fmt.Errorf("%w: invalid revoked device id %q", ErrInvalidRevocationRecord, revokedDeviceID)
	}
	if seq == 0 {
		return nil, fmt.Errorf("%w: seq must be greater than 0", ErrInvalidRevocationRecord)
	}

	tsStr := now.UTC().Format(time.RFC3339)
	challenge := BuildRevocationChallenge(id.DeviceID, revokedDeviceID, seq, tsStr)

	sig, err := id.Sign(challenge)
	if err != nil {
		return nil, fmt.Errorf("sign revocation record: %w", err)
	}

	return &RevocationRecord{
		RevokerDeviceID: id.DeviceID,
		RevokedDeviceID: revokedDeviceID,
		Seq:             seq,
		Timestamp:       tsStr,
		Signature:       hex.EncodeToString(sig),
	}, nil
}

// SignSelfTombstone creates and signs a new self-tombstone RevocationRecord announcing this device's own revocation (ADR 0010 §4.2).
func SignSelfTombstone(id *DeviceIdentity, seq uint64, now time.Time) (*RevocationRecord, error) {
	if id == nil {
		return nil, ErrInvalidIdentity
	}
	return SignRevocation(id, id.DeviceID, seq, now)
}

// Validate checks the structural syntax and field constraints of a RevocationRecord.
// Under ADR 0010 §4.2, self-tombstones (RevokerDeviceID == RevokedDeviceID) are structurally valid.
func (r *RevocationRecord) Validate() error {
	if r == nil {
		return ErrInvalidRevocationRecord
	}
	if !ValidateDeviceID(r.RevokerDeviceID) {
		return fmt.Errorf("%w: invalid revoker device id %q", ErrInvalidRevocationRecord, r.RevokerDeviceID)
	}
	if !ValidateDeviceID(r.RevokedDeviceID) {
		return fmt.Errorf("%w: invalid revoked device id %q", ErrInvalidRevocationRecord, r.RevokedDeviceID)
	}
	if r.Seq == 0 {
		return fmt.Errorf("%w: seq must be > 0", ErrInvalidRevocationRecord)
	}
	if _, err := time.Parse(time.RFC3339, r.Timestamp); err != nil {
		return fmt.Errorf("%w: invalid RFC3339 timestamp %q: %v", ErrInvalidRevocationRecord, r.Timestamp, err)
	}
	sigBytes, err := hex.DecodeString(r.Signature)
	if err != nil || len(sigBytes) != ed25519.SignatureSize {
		return fmt.Errorf("%w: signature must be 64 bytes hex", ErrInvalidRevocationRecord)
	}
	return nil
}

// VerifyRevocation validates the signature and timestamp constraints of a RevocationRecord against the revoker's public key.
func VerifyRevocation(rec *RevocationRecord, revokerPubKey ed25519.PublicKey, maxSkew time.Duration, now time.Time) error {
	if err := rec.Validate(); err != nil {
		return err
	}

	expectedRevokerID := DeriveDeviceID(revokerPubKey)
	if expectedRevokerID != rec.RevokerDeviceID {
		return fmt.Errorf("%w: revoker device ID %q does not match public key derived ID %q", ErrRevocationSignatureFailed, rec.RevokerDeviceID, expectedRevokerID)
	}

	ts, err := time.Parse(time.RFC3339, rec.Timestamp)
	if err != nil {
		return fmt.Errorf("%w: invalid timestamp", ErrInvalidRevocationRecord)
	}

	if maxSkew > 0 {
		skew := now.Sub(ts)
		if skew < -maxSkew {
			return fmt.Errorf("%w: timestamp is in the future beyond acceptable skew", ErrRevocationTimestampSkew)
		}
	}

	sigBytes, err := hex.DecodeString(rec.Signature)
	if err != nil || len(sigBytes) != ed25519.SignatureSize {
		return ErrRevocationSignatureFailed
	}

	challenge := BuildRevocationChallenge(rec.RevokerDeviceID, rec.RevokedDeviceID, rec.Seq, rec.Timestamp)
	if !VerifyDeviceSignature(revokerPubKey, challenge, sigBytes) {
		return ErrRevocationSignatureFailed
	}

	return nil
}
